packages/pcd/gpc-pcd/src/GPCPCDPackage.ts

```diff
@@ -0,0 +1,305 @@
+
+  if (!args.pod.value) {
+    throw new Error("No PODPCD value provided");
+  }
+  const podPCD = await PODPCDPackage.deserialize(args.pod.value.pcd);
+  if (!isPODPCD(podPCD)) {
+    throw new Error("Wrong PCD type provided for PODPCD");
+  }
+
+  const serializedIdentityPCD = args.identity.value?.pcd;
+  if (!serializedIdentityPCD) {
+    throw new Error("Cannot make proof: missing owner PCD");
+  }
+  const ownerSemaphorePCD =
+    serializedIdentityPCD !== undefined
+      ? await SemaphoreIdentityPCDPackage.deserialize(serializedIdentityPCD)
+      : undefined;
+
+  const externalNullifier =
+    args.externalNullifier.value !== undefined
+      ? ({
+          type: "string",
+          value: args.externalNullifier.value
+        } satisfies PODStringValue)
+      : undefined;
+  if (externalNullifier !== undefined && ownerSemaphorePCD === undefined) {
+    throw new Error("External nullifier requires an owner identity PCD.");
+  }
+  const watermark =
+    args.watermark.value !== undefined
+      ? ({
+          type: "string",
+          value: args.watermark.value
+        } satisfies PODStringValue)
+      : undefined;
+
+  return {
+    proofConfig,
+    proofInputs: {
+      pods: { pod0: podPCD.pod },
+      ...(ownerSemaphorePCD !== undefined
+        ? {
+            owner: {
+              semaphoreV3: ownerSemaphorePCD?.claim?.identity,
+              externalNullifier: externalNullifier
+            }
+          }
+        : {}),
+      watermark: watermark
+    }
+  };
+}
+
+/**
+ * Creates a new {@link GPCPCD} by generating an {@link GPCPCDProof}
+ * and deriving an {@link GPCPCDClaim} from the given {@link GPCPCDArgs}.
+ *
+ * This generates a ZK proof using the given config and inputs using a
+ * selected circuit from the supported family.
+ *
+ * @throws if the arguments are invalid
+ */
+export async function prove(args: GPCPCDArgs): Promise<GPCPCD> {
+  const zkArtifactPath = ensureInitialized();
+  const { proofConfig, proofInputs } = await checkProofArgs(args);
+  const id =
+    args.id !== undefined && typeof args.id.value === "string"
+      ? args.id.value
+      : uuid();
+
+  const { boundConfig, revealedClaims, proof } = await gpcProve(
+    proofConfig,
+    proofInputs,
+    zkArtifactPath
+  );
+
+  return new GPCPCD(
+    id,
+    {
+      config: boundConfig,
+      revealed: revealedClaims
+    },
+    { groth16Proof: proof }
+  );
+}
+
+/**
+ * Verifies a POD PCD by checking that its {@link GPCPCDClaim} corresponds to
+ * its {@link GPCPCDProof}.
+ *
+ * This confirms that the claimed configuration and revealed inputs match
+ * what was used to prove.
+ */
+export async function verify(pcd: GPCPCD): Promise<boolean> {
+  const zkArtifactPath = ensureInitialized();
+  try {
+    return gpcVerify(
+      pcd.proof.groth16Proof,
+      pcd.claim.config,
+      pcd.claim.revealed,
+      zkArtifactPath
+    );
+  } catch (e) {
+    console.error("Verifying invalid GPC proof data:", e);
```

Comment:
I don't think this needs changing now, but since this component can also be used on the server-side, logging to the console might not be the best way to record exceptions. This is something that @ichub has flagged up before, and I think he intends to review how we do logging across both client and server in order to ensure we're logging everything appropriately.

In the circumstances, it's better to log to the console than to log nothing, so this seems fine to me as-is, but also worth remembering to revisit at a suitable time in future.

Reply:
I'm ambivalent on the topic of whether verify should throw or return false on invalid input.  I thought I was following standard practice here, but it seems to be inconsistent.  EdDSAPCD has a catch block, but EdDSAEventTicketPCD doesn't.  Console logging was my best attempt to make sure the information at least goes somewhere.

I actually think letting verify() throw is better, particularly if deserializing already checked for most format issues (which it does for POD and GPC PCDs).  Callers can always catch if they want to treat an error the same as a validation failure.  I'll switch to that